[core] Make DashboardLayout navigation responsive

apedroferreira opened the pull request:
Closes https://github.com/mui/mui-toolpad/issues/3585

Make navigation menu open and close with a menu toggle in left side of the header in breakpoints under `md`.
Also centered logo + title in mobile screens.

**DashboardLayout docs**: https://deploy-preview-3750--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout


https://github.com/user-attachments/assets/7b417f09-3e01-41af-bd6b-7179faafae13

Janpot commented:
There are some problems in the docs:

* We may want to show off the navigation, even if the content is too small. e.g. like the current behavior
* Expanding the navigation behaves erratically

    https://github.com/mui/mui-toolpad/assets/2109932/bbc65cd9-d0b5-4a0c-a9fe-40687335c384

apedroferreira commented:
> There are some problems in the docs:

Oh thanks I actually hadn't checked the docs, assumed they'd be fine.

apedroferreira commented:
Added some fixes/improvements and addressed review feedback:
- Center logo/title in mobile header relative to the whole header width (instead of in the remaining space after the menu icon)
- Add `initialNavigationOpen` prop to start with mobile navigation open in the docs demos
- Use persistent drawer instead of temporary drawer - fixes the issue with unstyled giant icons in docs which I couldn't really find the cause for, and this way in some cases even with the mobile drawer open users can interact with the dashboard so that might be an advantage

Updated video in the PR description with this new behavior.

bharatkashyap reviewed (commented):
Looks good! About `initialNavigationOpen`, I would expect the name of the prop to match more closely with the behaviour of opening the _mobile_ navigation by default. 

Perhaps `mobileNavigationOpen`

Janpot commented on docs/translations/api-docs/dashboard-layout/dashboard-layout.json:
```diff
@@ -1,5 +1,10 @@
+  "propDescriptions": {
+    "children": { "description": "The content of the dashboard." },
+    "initialNavigationOpen": {
+      "description": "Whether the mobile navigation drawer should start open."
```
Should this be external API? Why would users need this?
Besides, can't we, for demo purposes just force the large screen behavior. Maybe change the breakpoints or something. But in a hidden way, users shouldn't copy+paste our hacks accidentally.

apedroferreira replied:
It should be internal API if possible, just not sure what the best way to do that is? Will try to find a solution.

Also wasn't sure if it's better to have the mobile or the desktop version on the docs, so I guess we can try to use the desktop one there or change the breakpoints.

apedroferreira commented:
> Looks good! About `initialNavigationOpen`, I would expect the name of the prop to match more closely with the behaviour of opening the _mobile_ navigation by default.
> 
> Perhaps `mobileNavigationOpen`

Just didn't want it to be too long but I guess we can go with `initialMobileNavigationOpen`

apedroferreira commented:
Addressed latest review feedback:
- `isDemoMode` internal prop (just simply marked it as internal) that enables a lower breakpoint for mobile view + mobile navigation starts open, for now.

apedroferreira commented:
> > Addressed latest review feedback:
> > 
> > * `isDemoMode` internal prop (just simply marked it as internal) that enables a lower breakpoint for mobile view + mobile navigation starts open, for now.
> 
> To me, adding an internal prop (`isDemoMode`) and all accompanying logic inside the component itself purely for demo purposes is bloat - why should a user intending on using that component pull in all that code?
> 
> I think we should try and obtain the behaviour we want on the docs by adding customisation over and above the component – perhaps a wrapper for the docs?

In my opinion this approach is fine as we're not adding a ton of props that do anything complex or add a lot of overhead, but we can discuss as a team on a solution for this type of scenario we can all agree on.

apedroferreira commented:
> > > Addressed latest review feedback:
> > > 
> > > * `isDemoMode` internal prop (just simply marked it as internal) that enables a lower breakpoint for mobile view + mobile navigation starts open, for now.
> > 
> > 
> > To me, adding an internal prop (`isDemoMode`) and all accompanying logic inside the component itself purely for demo purposes is bloat - why should a user intending on using that component pull in all that code?
> > I think we should try and obtain the behaviour we want on the docs by adding customisation over and above the component – perhaps a wrapper for the docs?
> 
> In my opinion this approach is fine as we're not adding a ton of props that do anything complex or add a lot of overhead, but we can discuss as a team on a solution for this type of scenario we can all agree on.
> 
> Another thing I could do but doesn't really tackle that issue is split the `isDemoMode` internal prop into 2 internal props: `initialMobileNavigationOpen` and `mobileBreakpoint`.

Oh, we also might be able to skip these demo-only props if the toggle for mobile/desktop view works well.

Janpot commented:
my idea was to provide a theme to the app provider that [redefined the breakpoint](https://mui.com/material-ui/customization/breakpoints/#custom-breakpoints) sizes in such a way that the Dashboardlayout shows the sidebar on smaller width. no new api required

apedroferreira commented:
> my idea was to provide a theme to the app provider that [redefined the breakpoint](https://mui.com/material-ui/customization/breakpoints/#custom-breakpoints) sizes in such a way that the Dashboardlayout shows the sidebar on smaller width. no new api required

Ah good idea, a theme might do it without being too hackish.

apedroferreira commented:
Solutions to the recent problems that have been raised:

- Added breakpoints to the theme in the docs examples so that the first view usually shown in docs does not have a mobile menu. The theme with these takes up a lot of lines and looks ugly in the current version of theming, but should be a lot smaller after the theming changes in https://github.com/mui/mui-toolpad/pull/3674 (no component overrides, will just need to extend the imported themes). We can wait for that PR to be merged first too.
- Reverted back to a "temporary" navigation drawer as shown in the re-updated video example above, as originally attempted, as I tried out the dashboard on my phone and this is a much better UX for mobile. This required a fix for the drawer in the docs, because the portal used by this navigation defaults to the document outside the iframes (as reported by Jan in this PR already). Fixed it the same way as the responsive drawer example in https://mui.com/material-ui/api/drawer/, and by adding a public `container` prop to the DashboardLayout component where the container for this portal can be set.

apedroferreira commented:
This one should be good for a review, the themes in the demos are cleaner now!
Check this comment for the latest changes: https://github.com/mui/mui-toolpad/pull/3750#issuecomment-2223598775

Janpot reviewed (approved):
Let's go!

apedroferreira commented:
Hi, the features in this PR are already live in the latest version of `@toolpad/core`.
Let us know if you find any issues or anything missing!

apedroferreira commented:
> > Dados #3585
> > Faça o menu de navegação abrir e fechar com um menu toggle no lado esquerdo do cabeçalho em breakpoints sob `md`. Também centralize o logotipo + título em telas móveis.
> > **Documentação do DashboardLayout** : https://deploy-preview-3750--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout
> > Gravação.de.tela.2024-07-19.at.14.33.41.mov
> 
> Hello, very good! Can you make the code available?

Hi, the features in this PR are already live in the latest version of @toolpad/core.
Let us know if you find any issues or anything missing!

FabioZan86 commented:
> > > Dados #3585
> > > Faça o menu de navegação abrir e fechar com um menu toggle no lado esquerdo do cabeçalho em breakpoints sob `md`. Também centralize o logotipo + título em telas móveis.
> > > **Documentação do DashboardLayout** : https://deploy-preview-3750--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout
> > > Gravação.de.tela.2024-07-19.at.14.33.41.mov
> > 
> > 
> > Hello, very good! Can you make the code available?
> 
> Hi, the features in this PR are already live in the latest version of `@toolpad/core`. Let us know if you find any issues or anything missing!

Hi...How can I customize the layout with the toolpad/core? Make the vertical bars in the collected style as in the photo? 
![collapsed sheels](https://github.com/user-attachments/assets/603f9a8e-99d7-46e7-9654-35a0efef1ee0)

apedroferreira commented:
> > > > Dados #3585
> > > > Faça o menu de navegação abrir e fechar com um menu toggle no lado esquerdo do cabeçalho em breakpoints sob `md`. Também centralize o logotipo + título em telas móveis.
> > > > **Documentação do DashboardLayout** : https://deploy-preview-3750--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout
> > > > Gravação.de.tela.2024-07-19.at.14.33.41.mov
> > > 
> > > 
> > > Hello, very good! Can you make the code available?
> > 
> > 
> > Hi, the features in this PR are already live in the latest version of `@toolpad/core`. Let us know if you find any issues or anything missing!
> 
> Hi...How can I customize the layout with the toolpad/core? Make the vertical bars in the collected style as in the photo? ![collapsed sheels](https://private-user-images.githubusercontent.com/90842406/378841549-603f9a8e-99d7-46e7-9654-35a0efef1ee0.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MDEzNTMsIm5iZiI6MTcyOTYwMTA1MywicGF0aCI6Ii85MDg0MjQwNi8zNzg4NDE1NDktNjAzZjlhOGUtOTlkNy00NmU3LTk2NTQtMzVhMGVmZWYxZWUwLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIyVDEyNDQxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViODE3NWY2YWE5YTdkOTQ1MTU3MTMzNTJmZWU3NGVkYzE1ZWZkMmExZjM3YzQzZjllY2Q1YmE4YTllNjk1OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K4-zfUkF3u_2NyY-6PvMvXYKWgtQBo58X-e7ebQYyOo)

So you want to have 2 columns on the left-side?
Right now it would be difficult to make the dashboard layout look exactly like that...
We will create slots that allow you to customize the layout a bit more, including one to override the navigation sidebar with any custom component.

I've created an issue in https://github.com/mui/toolpad/issues/4299, hopefully this feature will help!

FabioZan86 commented:
> > > > > Dados #3585
> > > > > Faça o menu de navegação abrir e fechar com um menu toggle no lado esquerdo do cabeçalho em breakpoints sob `md`. Também centralize o logotipo + título em telas móveis.
> > > > > **Documentação do DashboardLayout** : https://deploy-preview-3750--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout
> > > > > Gravação.de.tela.2024-07-19.at.14.33.41.mov
> > > > 
> > > > 
> > > > Hello, very good! Can you make the code available?
> > > 
> > > 
> > > Hi, the features in this PR are already live in the latest version of `@toolpad/core`. Let us know if you find any issues or anything missing!
> > 
> > 
> > Hi...How can I customize the layout with the toolpad/core? Make the vertical bars in the collected style as in the photo? ![collapsed sheels](https://private-user-images.githubusercontent.com/90842406/378841549-603f9a8e-99d7-46e7-9654-35a0efef1ee0.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MDEzNTMsIm5iZiI6MTcyOTYwMTA1MywicGF0aCI6Ii85MDg0MjQwNi8zNzg4NDE1NDktNjAzZjlhOGUtOTlkNy00NmU3LTk2NTQtMzVhMGVmZWYxZWUwLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIyVDEyNDQxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViODE3NWY2YWE5YTdkOTQ1MTU3MTMzNTJmZWU3NGVkYzE1ZWZkMmExZjM3YzQzZjllY2Q1YmE4YTllNjk1OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K4-zfUkF3u_2NyY-6PvMvXYKWgtQBo58X-e7ebQYyOo)
> 
> So you want to have 2 columns on the left-side? Right now it would be difficult to make the dashboard layout look exactly like that... We will create slots that allow you to customize the layout a bit more, including one to override the navigation sidebar with any custom component.
> 
> I've created an issue in #4299, hopefully this feature will help!

The toolpad doesn't make customizing more dynamic, right? So I can leave it aside and use old components, discarding the toolpad core. I want to create a double drawer...

apedroferreira commented:
> > > > > > Dados #3585
> > > > > > Faça o menu de navegação abrir e fechar com um menu toggle no lado esquerdo do cabeçalho em breakpoints sob `md`. Também centralize o logotipo + título em telas móveis.
> > > > > > **Documentação do DashboardLayout** : https://deploy-preview-3750--mui-toolpad-docs.netlify.app/toolpad/core/react-dashboard-layout
> > > > > > Gravação.de.tela.2024-07-19.at.14.33.41.mov
> > > > > 
> > > > > 
> > > > > Hello, very good! Can you make the code available?
> > > > 
> > > > 
> > > > Hi, the features in this PR are already live in the latest version of `@toolpad/core`. Let us know if you find any issues or anything missing!
> > > 
> > > 
> > > Hi...How can I customize the layout with the toolpad/core? Make the vertical bars in the collected style as in the photo? ![collapsed sheels](https://private-user-images.githubusercontent.com/90842406/378841549-603f9a8e-99d7-46e7-9654-35a0efef1ee0.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2MDEzNTMsIm5iZiI6MTcyOTYwMTA1MywicGF0aCI6Ii85MDg0MjQwNi8zNzg4NDE1NDktNjAzZjlhOGUtOTlkNy00NmU3LTk2NTQtMzVhMGVmZWYxZWUwLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIyVDEyNDQxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViODE3NWY2YWE5YTdkOTQ1MTU3MTMzNTJmZWU3NGVkYzE1ZWZkMmExZjM3YzQzZjllY2Q1YmE4YTllNjk1OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.K4-zfUkF3u_2NyY-6PvMvXYKWgtQBo58X-e7ebQYyOo)
> > 
> > 
> > So you want to have 2 columns on the left-side? Right now it would be difficult to make the dashboard layout look exactly like that... We will create slots that allow you to customize the layout a bit more, including one to override the navigation sidebar with any custom component.
> > I've created an issue in #4299, hopefully this feature will help!
> 
> The toolpad doesn't make customizing more dynamic, right? So I can leave it aside and use old components, discarding the toolpad core. I want to create a double drawer...

Depending on how specific the layout you want to create is, it might be difficult to adjust Toolpad Core to look exactly like it at the moment...
As we add more customization options it should eventually be possible, but not sure if simply being able to replace the current drawer and header with custom components will be enough for you, as that's one of the next upcoming features.
If it is, you can wait for that feature, otherwise maybe using standard Material UI components will be best for now.

apedroferreira commented:
> I urgently need to make this layout this way, separated by drawers like Instagram and responsive: ![insta](https://private-user-images.githubusercontent.com/90842406/379027851-94281a7c-c147-414f-8efc-b27aad22c8da.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3OTAxNTMsIm5iZiI6MTcyOTc4OTg1MywicGF0aCI6Ii85MDg0MjQwNi8zNzkwMjc4NTEtOTQyODFhN2MtYzE0Ny00MTRmLThlZmMtYjI3YWFkMjJjOGRhLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDE3MTA1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMjA0OGJiMTcyYzMzOWE2ZjBhYmZiZGViMmJjNWY1MDI5ZjVkMGE3NjA3YTFhYjQ4ODdlZGQwMGE5YTE3OTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P7tSCbC1mL0sXfZu4sogSd67fuLGG5mIFdnI1zPYAdA)

Will try to make it so you can override the header and drawer with your own custom components, and release it in v0.9.0 of `@toolpad/core`, but no exact timeframes yet.

I've also created an issue for supporting layouts with two navigation sidebars https://github.com/mui/toolpad/issues/4304 but I think it might not be our biggest priority for now.

FabioZan86 commented:
> > I urgently need to make this layout this way, separated by drawers like Instagram and responsive: ![insta](https://private-user-images.githubusercontent.com/90842406/379027851-94281a7c-c147-414f-8efc-b27aad22c8da.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3OTAxNTMsIm5iZiI6MTcyOTc4OTg1MywicGF0aCI6Ii85MDg0MjQwNi8zNzkwMjc4NTEtOTQyODFhN2MtYzE0Ny00MTRmLThlZmMtYjI3YWFkMjJjOGRhLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI0VDE3MTA1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMjA0OGJiMTcyYzMzOWE2ZjBhYmZiZGViMmJjNWY1MDI5ZjVkMGE3NjA3YTFhYjQ4ODdlZGQwMGE5YTE3OTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.P7tSCbC1mL0sXfZu4sogSd67fuLGG5mIFdnI1zPYAdA)
> 
> Will try to make it so you can override the header and drawer with your own custom components, and release it in v0.9.0 of `@toolpad/core`, but no exact timeframes yet.
> 
> I've also created an issue for supporting layouts with two navigation sidebars #4304 but I think it might not be our biggest priority for now.

Thanks! I'm building my screens with the @mui/core components!
